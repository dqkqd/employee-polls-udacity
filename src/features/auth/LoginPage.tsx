import { Box, Stack, Typography } from "@mui/material"
import LoginForm from "./LoginForm"

const LoginPage = () => {
  return (
    <Box justifyContent="center" display="flex" mt={10}>
      <Stack
        sx={{ width: "300px" }}
        spacing={2}
        component="form"
        style={{ textAlign: "center" }}
      >
        <Typography variant="h4" mt={2} fontWeight="bolder">
          Employee Polls
        </Typography>

        <img
          src="login-logo.jpg"
          alt="login logo"
          style={{ borderRadius: "100%" }}
        />

        <LoginForm />
      </Stack>
    </Box>
  )
}

export default LoginPage
