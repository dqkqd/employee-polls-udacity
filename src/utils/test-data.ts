import { EntityState } from "@reduxjs/toolkit"
import {
  getQuestions,
  getUsers,
  saveQuestion,
  saveQuestionAnswer,
  saveUser,
} from "../api"
import { AuthedUser, Question, User } from "../interfaces"
import { AnswerId } from "./../interfaces/index"

const fakeUsers = [
  {
    id: "@fake-user-1",
    name: "Fake User 1",
    password: "@fake-user-1-password",
    avatarURL: "https://picsum.photos/200",
  },
  {
    id: "@fake-user-2",
    name: "Fake User 2",
    password: "@fake-user-2-password",
    avatarURL: "https://picsum.photos/200",
  },
]

const fakeQuestions = [
  {
    optionOneText: "fake-question-1-option-one",
    optionTwoText: "fake-question-1-option-two",
    author: fakeUsers[0].id,
  },
  {
    optionOneText: "fake-question-2-option-one",
    optionTwoText: "fake-question-2-option-two",
    author: fakeUsers[1].id,
  },
  {
    optionOneText: "fake-question-3-option-one",
    optionTwoText: "fake-question-3-option-two",
    author: fakeUsers[1].id,
  },
]

const fakeQuestionsAnswers = [
  {
    userIndex: 0,
    questionIndex: 2, // fake-question-3
    answerId: "optionOne",
  },
]

export const initialAuth: AuthedUser = { ...fakeUsers[0], status: "success" }

const setupTestData = async () => {
  const addedUsers = await Promise.all(
    fakeUsers.map(async (user) => await saveUser(user)),
  )

  const addedQuestions = await Promise.all(
    fakeQuestions.map(async (question) => await saveQuestion(question)),
  )

  const addedAnswer = await Promise.all(
    fakeQuestionsAnswers.map(
      async ({ userIndex, questionIndex, answerId }) =>
        await saveQuestionAnswer({
          authedUser: addedUsers[userIndex].id,
          qid: addedQuestions[questionIndex].id,
          answerId: answerId as AnswerId,
        }),
    ),
  )

  if (!addedAnswer.every(Boolean)) {
    throw new Error("Some questions answer are not added")
  }

  const [allUsers, allQuestions] = await Promise.all([
    getUsers(),
    getQuestions(),
  ])
  const initialUsers: EntityState<User> = {
    ids: Object.keys(allUsers),
    entities: allUsers,
  }
  const initialQuestions: EntityState<Question> = {
    ids: Object.keys(allUsers),
    entities: allQuestions,
  }

  // need to fetch added users and added questions again because we added answer
  const addedUserIds = addedUsers.map((user) => user.id)
  const fetchAddedUsers = Object.values(allUsers).filter((user) =>
    addedUserIds.includes(user.id),
  )
  const addedQuestionIds = addedQuestions.map((question) => question.id)
  const fetchAddedQuestions = Object.values(allQuestions).filter((question) =>
    addedQuestionIds.includes(question.id),
  )

  return {
    initialUsers,
    initialQuestions,
    addedUsers: fetchAddedUsers,
    addedQuestions: fetchAddedQuestions,
  }
}

export const { initialUsers, initialQuestions, addedUsers, addedQuestions } =
  await setupTestData()
