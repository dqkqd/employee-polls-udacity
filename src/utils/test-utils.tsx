import type { PreloadedState } from "@reduxjs/toolkit"
import { configureStore } from "@reduxjs/toolkit"
import { render, type RenderOptions } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import React from "react"
import { Provider } from "react-redux"
import { createMemoryRouter, RouterProvider } from "react-router-dom"
import type { AppStore, RootState } from "../app/store"
import authSlice from "../features/auth/authSlice"
import questionReducer from "../features/questions/questionsSlice"
import usersSlice from "../features/users/usersSlice"
import { routesConfig } from "../routes"
import { initialUsers } from "./test-data"

// As a basic setup, import your same slice reducers
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  route?: string
  router?: ReturnType<typeof createMemoryRouter>
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    route = "/",
    preloadedState = {
      users: initialUsers,
    },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: {
        auth: authSlice,
        users: usersSlice,
        questions: questionReducer,
      },
      preloadedState,
    }),
    router = createMemoryRouter(routesConfig, {
      initialEntries: [route],
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper(): JSX.Element {
    return (
      <Provider store={store}>
        <RouterProvider router={router} />
      </Provider>
    )
  }

  window.history.pushState({}, "Test page", route)

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    user: userEvent.setup(),
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  }
}

export function renderDefault({
  route = "/",
  preloadedState = {
    users: initialUsers,
  },
  // Automatically create a store instance if no store was passed in
  store = configureStore({
    reducer: {
      auth: authSlice,
      users: usersSlice,
      questions: questionReducer,
    },
    preloadedState,
  }),
  router = createMemoryRouter(routesConfig, {
    initialEntries: [route],
  }),
  ...renderOptions
}: ExtendedRenderOptions = {}) {
  return renderWithProviders(<></>, {
    route,
    preloadedState,
    store,
    router,
    ...renderOptions,
  })
}
