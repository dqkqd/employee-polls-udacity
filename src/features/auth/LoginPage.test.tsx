import { screen } from "@testing-library/react"
import { expect, it } from "vitest"
import { renderWithNoRoutes } from "../../utils/test-utils"
import LoginPage from "./LoginPage"

it("Render", () => {
  const { baseElement } = renderWithNoRoutes(<LoginPage />)
  expect(baseElement).toMatchSnapshot()

  expect(screen.getByLabelText("log-in-page-title")).toHaveTextContent("Log In")
  expect(screen.getByLabelText("Employee ID")).toBeInTheDocument()
  expect(screen.getByLabelText("Password")).toBeInTheDocument()
  expect(screen.getByRole("button", { name: "Log In" })).toBeInTheDocument()
})
