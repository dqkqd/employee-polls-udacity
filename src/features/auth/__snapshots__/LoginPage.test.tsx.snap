// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render 1`] = `
<body>
  <div>
    <div
      class="MuiBox-root css-elt1hl"
    >
      <form
        class="MuiStack-root css-cqw1hc-MuiStack-root"
        style="text-align: center;"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 css-1fnef21-MuiTypography-root"
        >
          Employee Polls
        </h4>
        <img
          alt="login logo"
          src="/src/login-logo.jpg"
          style="border-radius: 100%; height: 300px; width: 300px; margin: auto;"
        />
        <h5
          aria-label="log-in-page-title"
          class="MuiTypography-root MuiTypography-h5 css-1e53d7t-MuiTypography-root"
        >
          Log In
        </h5>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Employee ID
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Employee ID
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              type="password"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="VisibilityIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="submit"
        >
          Log In
        </button>
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          Don't have an account?
          <a
            href="/signup"
            style="margin: 5px;"
          >
            Sign up
          </a>
          or
          <a
            href="/login/users"
            style="margin: 5px;"
          >
            Login using a sample account
          </a>
        </p>
      </form>
    </div>
  </div>
</body>
`;
